File: ocaml/src/crypto.ml
(* SHA-256 and HMAC-SHA256, hand-rolled.

   OCaml's distribution has no cryptographic digest but MD5, so the two
   primitives SigV4 is built from are written out here, straight from FIPS
   180-4 and RFC 2104.

   The one dependency this port takes is OpenSSL, and it is taken for
   TRANSPORT only (src/tls_stubs.c). Reaching into libcrypto for a digest
   would widen that exception from "cryptographic transport is not
   hand-rolled" to "cryptography is not hand-rolled", which is not the rule.
   Rust is the worked precedent: `ring` is already inside rustls's closure
   and `rust/src/crypto.rs` still carries both primitives.

   Correctness is not argued from inspection: a SigV4 signature is a chain
   of these two functions, so a single wrong bit anywhere fails the five
   known-answer vectors in the shared spec - one of which is AWS's own
   published `get-vanilla`. *)

let mask = 0xFFFFFFFF

let k =
  [| 0x428a2f98; 0x71374491; 0xb5c0fbcf; 0xe9b5dba5; 0x3956c25b; 0x59f111f1; 0x923f82a4;
     0xab1c5ed5; 0xd807aa98; 0x12835b01; 0x243185be; 0x550c7dc3; 0x72be5d74; 0x80deb1fe;
     0x9bdc06a7; 0xc19bf174; 0xe49b69c1; 0xefbe4786; 0x0fc19dc6; 0x240ca1cc; 0x2de92c6f;
     0x4a7484aa; 0x5cb0a9dc; 0x76f988da; 0x983e5152; 0xa831c66d; 0xb00327c8; 0xbf597fc7;
     0xc6e00bf3; 0xd5a79147; 0x06ca6351; 0x14292967; 0x27b70a85; 0x2e1b2138; 0x4d2c6dfc;
     0x53380d13; 0x650a7354; 0x766a0abb; 0x81c2c92e; 0x92722c85; 0xa2bfe8a1; 0xa81a664b;
     0xc24b8b70; 0xc76c51a3; 0xd192e819; 0xd6990624; 0xf40e3585; 0x106aa070; 0x19a4c116;
     0x1e376c08; 0x2748774c; 0x34b0bcb5; 0x391c0cb3; 0x4ed8aa4a; 0x5b9cca4f; 0x682e6ff3;
     0x748f82ee; 0x78a5636f; 0x84c87814; 0x8cc70208; 0x90befffa; 0xa4506ceb; 0xbef9a3f7;
     0xc67178f2 |]

let initial = [| 0x6a09e667; 0xbb67ae85; 0x3c6ef372; 0xa54ff53a; 0x510e527f; 0x9b05688c;
                 0x1f83d9ab; 0x5be0cd19 |]

let rotr x n = ((x lsr n) lor (x lsl (32 - n))) land mask
let shr x n = (x lsr n) land mask

(* The digest of a message, as 32 raw bytes. *)
let sha256 (message : string) : string =
  let len = String.length message in

  (* 0x80, then zeros until the length lands 8 short of a block, then the
     message length in BITS, big-endian. *)
  let padded = Buffer.create (len + 72) in
  Buffer.add_string padded message;
  Buffer.add_char padded '\x80';
  while (Buffer.length padded + 8) mod 64 <> 0 do
    Buffer.add_char padded '\x00'
  done;
  let bits = len * 8 in
  for shift = 7 downto 0 do
    Buffer.add_char padded (Char.chr ((bits lsr (shift * 8)) land 0xff))
  done;

  let block = Buffer.contents padded in
  let state = Array.copy initial in
  let w = Array.make 64 0 in

  let blocks = String.length block / 64 in

  for index = 0 to blocks - 1 do
    let base = index * 64 in

    for step = 0 to 15 do
      let at = base + (step * 4) in
      w.(step) <-
        (Char.code block.[at] lsl 24)
        lor (Char.code block.[at + 1] lsl 16)
        lor (Char.code block.[at + 2] lsl 8)
        lor Char.code block.[at + 3]
    done;

    for step = 16 to 63 do
      let s0 = rotr w.(step - 15) 7 lxor rotr w.(step - 15) 18 lxor shr w.(step - 15) 3 in
      let s1 = rotr w.(step - 2) 17 lxor rotr w.(step - 2) 19 lxor shr w.(step - 2) 10 in
      w.(step) <- (w.(step - 16) + s0 + w.(step - 7) + s1) land mask
    done;

    let a = ref state.(0) and b = ref state.(1) and c = ref state.(2) and d = ref state.(3) in
    let e = ref state.(4) and f = ref state.(5) and g = ref state.(6) and h = ref state.(7) in

    for step = 0 to 63 do
      let s1 = rotr !e 6 lxor rotr !e 11 lxor rotr !e 25 in
      let ch = (!e land !f) lxor (lnot !e land !g land mask) in
      let temp1 = (!h + s1 + ch + k.(step) + w.(step)) land mask in
      let s0 = rotr !a 2 lxor rotr !a 13 lxor rotr !a 22 in
      let maj = (!a land !b) lxor (!a land !c) lxor (!b land !c) in
      let temp2 = (s0 + maj) land mask in

      h := !g;
      g := !f;
      f := !e;
      e := (!d + temp1) land mask;
      d := !c;
      c := !b;
      b := !a;
      a := (temp1 + temp2) land mask
    done;

    state.(0) <- (state.(0) + !a) land mask;
    state.(1) <- (state.(1) + !b) land mask;
    state.(2) <- (state.(2) + !c) land mask;
    state.(3) <- (state.(3) + !d) land mask;
    state.(4) <- (state.(4) + !e) land mask;
    state.(5) <- (state.(5) + !f) land mask;
    state.(6) <- (state.(6) + !g) land mask;
    state.(7) <- (state.(7) + !h) land mask
  done;

  let out = Bytes.create 32 in
  Array.iteri
    (fun index word ->
      Bytes.set out (index * 4) (Char.chr ((word lsr 24) land 0xff));
      Bytes.set out ((index * 4) + 1) (Char.chr ((word lsr 16) land 0xff));
      Bytes.set out ((index * 4) + 2) (Char.chr ((word lsr 8) land 0xff));
      Bytes.set out ((index * 4) + 3) (Char.chr (word land 0xff)))
    state;

  Bytes.to_string out

(* Lowercase hex, two digits a byte. Uppercase hex is wanted in exactly one
   place in this library, and that is percent-escaping. *)
let hex (raw : string) : string =
  let out = Buffer.create (String.length raw * 2) in
  String.iter (fun ch -> Buffer.add_string out (Printf.sprintf "%02x" (Char.code ch))) raw;
  Buffer.contents out

let sha256hex (message : string) : string = hex (sha256 message)

(* RFC 2104, block size 64. Argument order is (key, data) in every port;
   two standard libraries take it the other way round. *)
let hmac (key : string) (data : string) : string =
  let block = 64 in
  let shortened = if String.length key > block then sha256 key else key in
  let padkey = Bytes.make block '\x00' in
  Bytes.blit_string shortened 0 padkey 0 (String.length shortened);

  let inner = Bytes.create block and outer = Bytes.create block in
  for index = 0 to block - 1 do
    let byte = Char.code (Bytes.get padkey index) in
    Bytes.set inner index (Char.chr (byte lxor 0x36));
    Bytes.set outer index (Char.chr (byte lxor 0x5c))
  done;

  sha256 (Bytes.to_string outer ^ sha256 (Bytes.to_string inner ^ data))

File: haskell/plugins/Crypto.hs
-- | SHA-256 and HMAC-SHA256, hand-rolled.
--
-- GHC ships no hash functions, and SigV4 signing needs exactly these two.
-- The TLS exception covers cryptographic TRANSPORT and nothing else, so
-- calling libcrypto's EVP_Digest here - which is linked in, and would work
-- - would be a rule violation rather than a shortcut. Rust is the worked
-- precedent: `ring` is already inside rustls's closure and the Rust port
-- still hand-rolls both.
--
-- SHA-256 is FIPS 180-4 straight from the standard; HMAC is RFC 2104 over
-- it. Both are proved by the spec's sigv4 known-answer cases, AWS's own
-- published `get-vanilla` vector among them: a signature is a chain of
-- these primitives, so one wrong bit anywhere fails there.

module Crypto
  ( hmacsha256,
    sha256,
    sha256hex,
  )
where

import Bytes (hex, utf8encode)
import Data.Bits (complement, rotateR, shiftL, shiftR, xor, (.&.), (.|.))
import qualified Data.ByteString as B
import Data.List (foldl')
import Data.Word (Word32, Word64, Word8)

-- | The round constants: the fractional parts of the cube roots of the
-- first 64 primes.
constants :: [Word32]
constants =
  [ 0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1,
    0x923f82a4, 0xab1c5ed5, 0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
    0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174, 0xe49b69c1, 0xefbe4786,
    0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
    0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147,
    0x06ca6351, 0x14292967, 0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
    0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85, 0xa2bfe8a1, 0xa81a664b,
    0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
    0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a,
    0x5b9cca4f, 0x682e6ff3, 0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
    0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2
  ]

-- | The initial hash: the fractional parts of the square roots of the
-- first eight primes.
initial :: [Word32]
initial =
  [0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, 0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19]

-- | The SHA-256 digest of a byte string.
sha256 :: B.ByteString -> B.ByteString
sha256 message = B.pack (concatMap bebytes (foldl' block initial (chunks padded)))
  where
    bits = 8 * fromIntegral (B.length message) :: Word64

    -- 0x80, zeros up to 56 bytes into the last block, then the message
    -- length in BITS, big-endian.
    padded =
      B.concat
        [ message,
          B.singleton 0x80,
          B.replicate zeros 0,
          B.pack [fromIntegral (bits `shiftR` shift) | shift <- [56, 48 .. 0]]
        ]

    zeros = (56 - (B.length message + 1) `mod` 64 + 64) `mod` 64

    chunks raw
      | B.null raw = []
      | otherwise = B.take 64 raw : chunks (B.drop 64 raw)

-- | Bytes as lowercase hex, the form SigV4 embeds.
sha256hex :: String -> String
sha256hex = hex . sha256 . utf8encode

-- | One 64-byte block folded into the running hash.
block :: [Word32] -> B.ByteString -> [Word32]
block hash raw = zipWith (+) hash (compress hash schedule)
  where
    first16 = [word (B.take 4 (B.drop (4 * slot) raw)) | slot <- [0 .. 15]]

    -- The message schedule, defined in terms of itself: laziness gives the
    -- recurrence directly, and taking 64 of it forces exactly what is used.
    schedule = take 64 grown
    grown =
      first16
        ++ [ small0 (grown !! (slot - 15))
               + grown !! (slot - 16)
               + grown !! (slot - 7)
               + small1 (grown !! (slot - 2))
             | slot <- [16 .. 63]
           ]

    small0 value = rotateR value 7 `xor` rotateR value 18 `xor` (value `shiftR` 3)
    small1 value = rotateR value 17 `xor` rotateR value 19 `xor` (value `shiftR` 10)

    word bytes = foldl' (\at byte -> (at `shiftL` 8) .|. fromIntegral byte) 0 (B.unpack bytes)

compress :: [Word32] -> [Word32] -> [Word32]
compress hash schedule = foldl' round8 hash (zip constants schedule)
  where
    round8 [a, b, c, d, e, f, g, h] (constant, word) =
      [temp1 + temp2, a, b, c, d + temp1, e, f, g]
      where
        big1 = rotateR e 6 `xor` rotateR e 11 `xor` rotateR e 25
        choose = (e .&. f) `xor` (complement e .&. g)
        temp1 = h + big1 + choose + constant + word
        big0 = rotateR a 2 `xor` rotateR a 13 `xor` rotateR a 22
        majority = (a .&. b) `xor` (a .&. c) `xor` (b .&. c)
        temp2 = big0 + majority
    round8 other _ = other

bebytes :: Word32 -> [Word8]
bebytes value = [fromIntegral (value `shiftR` shift) | shift <- [24, 16, 8, 0]]

-- | HMAC-SHA256 (RFC 2104): the primitive SigV4's key derivation chains.
--
-- The argument order is (key, data), which is the convention every port
-- uses; some standard libraries take them the other way round.
hmacsha256 :: B.ByteString -> B.ByteString -> B.ByteString
hmacsha256 key payload = sha256 (B.append (mask 0x5c) (sha256 (B.append (mask 0x36) payload)))
  where
    -- A key longer than the 64-byte block is hashed down; a shorter one is
    -- zero-padded up.
    shortened = if 64 < B.length key then sha256 key else key
    block64 = B.append shortened (B.replicate (64 - B.length shortened) 0)
    mask pad = B.map (`xor` pad) block64
